src/main/java/com/spotify/google/cloud/pubsub/client/Puller.java

```diff
@@ -146,25 +148,32 @@ public long pullIntervalMillis() {
 
   private void pull() {
     while (outstandingRequests.get() < concurrency &&
-           outstandingMessages.get() < maxOutstandingMessages) {
+           outstandingMessages.get() < maxOutstandingMessages &&
+           hasData.get()) {
       pullBatch();
     }
   }
 
   private void pullBatch() {
     outstandingRequests.incrementAndGet();
 
-    pubsub.pull(project, subscription, false, batchSize)
+    pubsub.pull(project, subscription, true, batchSize)
         .whenComplete((messages, ex) -> {
 
           outstandingRequests.decrementAndGet();
```

Comment:
You can control concurrency by delaying decrementing of this variable in case you get an answer which is <batchSize.